### ISSUE TITLE
Strange behavior of Validation score during training when Forward(is_train=False)

### ISSUE THREAD

**Issue by zhreshold** [dmlc/mxnet#3374]
I encountered a very strange problem when training a detection network.
Training accuracy increase fine, however, validation is always like 0.01.

Tried debugging into the code, found that during training the output of network (which is then plugged into eval_metric) is good, however, during validation, the output is not correct. To be more specific, the output of the network is mx.symbol.Group([cls_prob, loc_loss, cls_label]) where:
cls_prob: generated class softmax prob
loc_loss: bounding box regression loss
cls_label: generated class target according to original ground-truth, (I have multiple class targets per image)

So I tried to print out cls_label, found that it is not changing at all. All validation cls_label is the same as the last output in training.

Finally I tried change the forward mode (is_train=False) in base_module.score to is_train=True 
https://github.com/dmlc/mxnet/blob/master/python/mxnet/module/base_module.py#L164
Everything is fine now. Validation accuracy  as good as training now.

I don't know what's going wrong when is_train=False, I suspect a bug in retrieving output.
Any help?

**Comment by piiswrong** [comment]
which layer outputs cls_label?

**Comment by zhreshold** [comment]
Ok, I will post the network here:

    data = mx.symbol.Variable(name="data")
    label = mx.symbol.Variable(name="label")

    # group 1
    conv1_1 = mx.symbol.Convolution(
        data=data, kernel=(3, 3), pad=(1, 1), num_filter=64, name="conv1_1")
    relu1_1 = mx.symbol.Activation(data=conv1_1, act_type="relu", name="relu1_1")
    conv1_2 = mx.symbol.Convolution(
        data=relu1_1, kernel=(3, 3), pad=(1, 1), num_filter=64, name="conv1_2")
    relu1_2 = mx.symbol.Activation(data=conv1_2, act_type="relu", name="relu1_2")
    pool1 = mx.symbol.Pooling(
        data=relu1_2, pool_type="max", kernel=(2, 2), stride=(2, 2), name="pool1")
    # group 2
    conv2_1 = mx.symbol.Convolution(
        data=pool1, kernel=(3, 3), pad=(1, 1), num_filter=128, name="conv2_1")
    relu2_1 = mx.symbol.Activation(data=conv2_1, act_type="relu", name="relu2_1")
    conv2_2 = mx.symbol.Convolution(
        data=relu2_1, kernel=(3, 3), pad=(1, 1), num_filter=128, name="conv2_2")
    relu2_2 = mx.symbol.Activation(data=conv2_2, act_type="relu", name="relu2_2")
    pool2 = mx.symbol.Pooling(
        data=relu2_2, pool_type="max", kernel=(2, 2), stride=(2, 2), name="pool2")
    # group 3
    conv3_1 = mx.symbol.Convolution(
        data=pool2, kernel=(3, 3), pad=(1, 1), num_filter=256, name="conv3_1")
    relu3_1 = mx.symbol.Activation(data=conv3_1, act_type="relu", name="relu3_1")
    conv3_2 = mx.symbol.Convolution(
        data=relu3_1, kernel=(3, 3), pad=(1, 1), num_filter=256, name="conv3_2")
    relu3_2 = mx.symbol.Activation(data=conv3_2, act_type="relu", name="relu3_2")
    conv3_3 = mx.symbol.Convolution(
        data=relu3_2, kernel=(3, 3), pad=(1, 1), num_filter=256, name="conv3_3")
    relu3_3 = mx.symbol.Activation(data=conv3_3, act_type="relu", name="relu3_3")
    pool3 = mx.symbol.Pooling(
        data=relu3_3, pool_type="max", kernel=(2, 2), stride=(2, 2), \
        pad=(1,1), name="pool3")
    # group 4
    conv4_1 = mx.symbol.Convolution(
        data=pool3, kernel=(3, 3), pad=(1, 1), num_filter=512, name="conv4_1")
    relu4_1 = mx.symbol.Activation(data=conv4_1, act_type="relu", name="relu4_1")
    conv4_2 = mx.symbol.Convolution(
        data=relu4_1, kernel=(3, 3), pad=(1, 1), num_filter=512, name="conv4_2")
    relu4_2 = mx.symbol.Activation(data=conv4_2, act_type="relu", name="relu4_2")
    conv4_3 = mx.symbol.Convolution(
        data=relu4_2, kernel=(3, 3), pad=(1, 1), num_filter=512, name="conv4_3")
    relu4_3 = mx.symbol.Activation(data=conv4_3, act_type="relu", name="relu4_3")
    pool4 = mx.symbol.Pooling(
        data=relu4_3, pool_type="max", kernel=(2, 2), stride=(2, 2), name="pool4")
    # group 5
    conv5_1 = mx.symbol.Convolution(
        data=pool4, kernel=(3, 3), pad=(1, 1), num_filter=512, name="conv5_1")
    relu5_1 = mx.symbol.Activation(data=conv5_1, act_type="relu", name="relu5_1")
    conv5_2 = mx.symbol.Convolution(
        data=relu5_1, kernel=(3, 3), pad=(1, 1), num_filter=512, name="conv5_2")
    relu5_2 = mx.symbol.Activation(data=conv5_2, act_type="relu", name="relu5_2")
    conv5_3 = mx.symbol.Convolution(
        data=relu5_2, kernel=(3, 3), pad=(1, 1), num_filter=512, name="conv5_3")
    relu5_3 = mx.symbol.Activation(data=conv5_3, act_type="relu", name="relu5_3")
    pool5 = mx.symbol.Pooling(
        data=relu5_3, pool_type="max", kernel=(3, 3), stride=(1, 1),
        pad=(1,1), name="pool5")
    # group 6
    conv6 = mx.symbol.Convolution(
        data=pool5, kernel=(3, 3), pad=(6, 6), dilate=(6, 6),
        num_filter=1024, name="conv6")
    relu6 = mx.symbol.Activation(data=conv6, act_type="relu", name="relu6")
    # drop6 = mx.symbol.Dropout(data=relu6, p=0.5, name="drop6")
    # group 7
    conv7 = mx.symbol.Convolution(
        data=relu6, kernel=(1, 1), pad=(0, 0), num_filter=1024, name="conv7")
    relu7 = mx.symbol.Activation(data=conv7, act_type="relu", name="relu7")
    # drop7 = mx.symbol.Dropout(data=relu7, p=0.5, name="drop7")

    ### ssd extra layers ###
    relu8_1 = conv_act_layer(relu7, "8_1", 256, kernel=(1,1), pad=(0,0), \
        stride=(1,1), act_type="relu", use_batchnorm=False)
    relu8_2 = conv_act_layer(relu8_1, "8_2", 512, kernel=(3,3), pad=(1,1), \
        stride=(2,2), act_type="relu", use_batchnorm=False)
    relu9_1 = conv_act_layer(relu8_2, "9_1", 128, kernel=(1,1), pad=(0,0), \
        stride=(1,1), act_type="relu", use_batchnorm=False)
    relu9_2 = conv_act_layer(relu9_1, "9_2", 256, kernel=(3,3), pad=(1,1), \
        stride=(2,2), act_type="relu", use_batchnorm=False)
    relu10_1 = conv_act_layer(relu9_2, "10_1", 128, kernel=(1,1), pad=(0,0), \
        stride=(1,1), act_type="relu", use_batchnorm=False)
    relu10_2 = conv_act_layer(relu10_1, "10_2", 256, kernel=(3,3), pad=(1,1), \
        stride=(2,2), act_type="relu", use_batchnorm=False)
    # global Pooling
    pool10 = mx.symbol.Pooling(data=relu10_2, pool_type="avg",
        global_pool=True, kernel=(1,1), name='pool10')

    # specific parameters for VGG16 network
    from_layers = [relu4_3, relu7, relu8_2, relu9_2, relu10_2, pool10]
    sizes = [[.1], [.2,.276], [.38, .461], [.56, .644], [.74, .825], [.92, 1.01]]
    ratios = [[1,2,.5], [1,2,3,.5,1./3], [1,2,3,.5,1./3], [1,2,3,.5,1./3], \
        [1,2,3,.5,1./3], [1,2,3,.5,1./3]]
    normalizations = [20, -1, -1, -1, -1, -1]
    # normalizations = [1, 1, 1, 1, 1, 1]

    loc_preds, cls_preds, anchor_boxes = multibox_layer(from_layers, \
        num_classes, sizes=sizes, ratios=ratios, normalization=normalizations, \
        clip=True, interm_layer=0)

    tmp = mx.symbol.MultiBoxTarget(
        *[anchor_boxes, label, cls_preds, loc_preds], overlap_threshold=.5, \
        ignore_label=-1, negative_mining_ratio=3, minimum_negative_samples=0, \
        negative_mining_thresh=.5, variances=(0.1, 0.1, 0.2, 0.2),
        name="multibox_target")
    loc_target = tmp[0]
    loc_target_mask = tmp[1]
    cls_target = tmp[2]

    cls_prob = mx.symbol.SoftmaxOutput(data=cls_preds, label=cls_target, \
        ignore_label=-1, use_ignore=True, grad_scale=1., multi_output=True, \
        normalization='valid', name="cls_prob")
    loc_loss_ = mx.symbol.smooth_l1(name="loc_loss_", \
        data=loc_target_mask * (loc_preds - loc_target), scalar=1.0)
    loc_loss = mx.symbol.MakeLoss(*[loc_loss_, loc_target_mask], \
        grad_scale=1., normalization='valid', name="loc_loss")
    # loc_loss = mx.symbol.LogisticRegressionOutput(data=loc_preds * loc_target_mask,
    #     label=loc_target * loc_target_mask, grad_scale=1, name="loc_loss")

    # monitoring training status
    cls_label = mx.symbol.MakeLoss(data=cls_target, grad_scale=0, name="cls_label")

    # group output
    # out = mx.symbol.Group([cls_prob, loc_loss])
    out = mx.symbol.Group([cls_prob, loc_loss, cls_label])
    return out

So cls_label is the third output of MultiBoxTarget layer, MultiBoxTarget is the new layer I wrote as in here
https://github.com/zhreshold/mxnet-ssd/blob/master/src/operator/multibox_target-inl.h

Do you need any other info @piiswrong ? Basically I have committed temporary work I wrote in this fork: 
https://github.com/zhreshold/mxnet-ssd

**Comment by piiswrong** [comment]
is if changing this line https://github.com/dmlc/mxnet/blob/master/python/mxnet/module/executor_group.py#L215 to `is True` helps.

@pluskid This is probably a bug, we may want to load the labels whenever it's provided, because when you use MakeLoss to calculate error in the graph instead of in metric, we want to load labels for testing

**Comment by pluskid** [comment]
I have two proposals:

1. For this kind of applications, maybe we can just treat label as part of data. In other words, provide both label and data from the data, and let label be empty.

2. Add an extra flag like `always_load_labels` to force the modules to load label even when not training.

**Comment by zhreshold** [comment]
Thanks for the suggestion. Now I understand the behaviour why label is not correctly updated.

**Comment by piiswrong** [comment]
@pluskid Why not just load labels whenever they are in the batch? regardless of training or testing? I can't think of a situation where this would cause problems

**Comment by zhreshold** [comment]
Agree with @piiswrong, since the presence of label is checked in forward, there's no need to check is_train before that.
The choice of whether or not providing label is then handled by users.
Simply comment this line should be fine.
https://github.com/dmlc/mxnet/blob/master/python/mxnet/module/executor_group.py#L215
@pluskid

**Comment by pluskid** [comment]
@piiswrong @zhreshold I was vaguely worrying about some other logic might break, but I could not recall what I had in mind now. I think what you said makes sense. Could any of you make a quick PR to fix this? Thanks a lot!

**Comment by zhreshold** [comment]
Submitted https://github.com/dmlc/mxnet/pull/3394
